cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,440 @@
+  // |variant_| member to contain |status| and sets |has_value_| to indicate
+  // that |this| does not have a value. Destroys the existing |variant_| member.
+  template <class U>
+  void OverwriteValueWithStatus(U&& status) {
+#ifndef NDEBUG
+    if (ARROW_PREDICT_FALSE(!ok())) {
+      internal::DieWithMessage("OverwriteValueWithStatus called on with existing error.");
+    }
+#endif
+    variant_.value_.~T();
+    // Re-use the memory from variant when constructing
+    new (&variant_) variant(std::forward<U>(status));
+    has_value_ = false;
+  }
+
+  // Resets the |variant_| member to contain the |value| and sets |has_value_|
+  // to indicate that the ErrorOr object has a value. Destroys the existing
+  // |variant_| member.
+  template <class U>
+  void AssignValue(U&& value) {
+    if (ok()) {
+      // We cannot assume that T is move-assignable.
+      variant_.value_.~T();
+    } else {
+      variant_.status_.~Status();
+    }
+    // Re-use the memory from variant when constructing
+    new (&variant_) variant(std::forward<U>(value));
+    has_value_ = true;
+  }
+
+  // Use custom variant instead of c++ standard, to avoid header polution.
+  union variant {
+    // A non-OK status.
+    Status status_;
+
+    // An element of type T.
+    T value_;
+
+    variant() {}
+
+    variant(const Status& status) : status_(status) {}
+
+    variant(Status&& status) : status_(std::move(status)) {}
+
+    template <typename U, typename E = typename std::enable_if<
+                              std::is_constructible<T, U>::value &&
+                              std::is_convertible<U, T>::value>::type>
+    variant(U&& value) : value_(std::forward<U>(value)) {}
+
+    // This destructor must be explicitly defined because it is deleted due to
+    // the variant type having non-static data members with non-trivial
+    // destructors.
+    ~variant() {}
+  };
+
+  // One of: a non-OK status or an element of type T.
+  variant variant_;
+
+  // Indicates the active member of the variant_ member.
+  //
+  // A value of true indicates that value_ is the active member of variant_.
+  //
+  // A value of false indicates that status_ is the active member of variant_.
+  bool has_value_;
+};
+
+#define ASSIGN_OR_RAISE_IMPL(status_name, lhs, rexpr) \
+  auto status_name = (rexpr);                         \
+  ARROW_RETURN_NOT_OK(status_name.status());          \
+  lhs = std::move(status_name).ValueOrDie();
+
+#define ASSIGN_OR_RAISE_NAME(x, y) ARROW_CONCAT(x, y)
+
+// Executes an expression that returns a ErrorOr, extracting its value
+// into the variable defined by lhs (or returning on error).
+//
+// Example: Assigning to an existing value
+//   ValueType value;
+//   ASSIGN_OR_RAISE(value, MaybeGetValue(arg));
+//
+// WARNING: ASSIGN_OR_RAISE expands into multiple statements; it cannot be used
+//  in a single statement (e.g. as the body of an if statement without {})!
+#define ASSIGN_OR_RAISE(lhs, rexpr) \
```

Comment:
By the way, macros exposed in headers should probably begin with ARROW_.

Reply:
fixed, add an internal header that redefines it to something shorter.